p2p/src/message_network.rs

```diff
@@ -86,6 +86,161 @@ impl_consensus_encoding!(
     relay
 );
 
+/// A bitcoin user agent defined by BIP-14. The user agent is sent in the version message when a
+/// connection between two peers is established. It is intended to advertise client software in a
+/// well-defined format.
+///
+/// ref: <https://github.com/bitcoin/bips/blob/master/bip-0014.mediawiki>
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct UserAgent {
+    user_agent: String,
+}
+
+impl_consensus_encoding!(UserAgent, user_agent);
+
+impl UserAgent {
+    /// Build a new user agent from the lowest level client software. For example: `Satoshi` is
+    /// used by Bitcoin Core.
+    ///
+    /// # Panics
+    ///
+    /// If the client name contains one of: `/ ( ) :`
+    pub fn new<S: AsRef<str>>(client_name: S, client_version: UserAgentVersion) -> Self {
+        let parsed_name = client_name.as_ref();
+        if parsed_name.chars().any(|c| matches!(c, '/' | '(' | ')' | ':')) {
+            panic!("client name cannot contain: / ( ) :");
+        }
+        let version = client_version.to_string();
+        let agent = format!("/{parsed_name}:{version}/");
+        Self { user_agent: agent }
+    }
+
+    /// Build a user agent, ignoring BIP-14 recommendations.
+    pub fn from_nonstandard<S: ToString>(agent: S) -> Self {
+        Self {
+            user_agent: agent.to_string()
+        }
+    }
+
+    /// Add a client to the user agent string. Examples may include the name of a wallet software.
+    ///
+    /// # Panics
+    ///
+    /// If the client name contains one of: `/ ( ) :`
+    #[must_use]
+    pub fn add_client<S: AsRef<str>>(mut self, client_name: S, client_version: UserAgentVersion) -> Self {
+        let parsed_name = client_name.as_ref();
+        if parsed_name.chars().any(|c| matches!(c, '/' | '(' | ')' | ':')) {
+            panic!("client name cannot contain: / ( ) :");
```

Comment:
Why are we panicing here? Presumably its reasonable to think that this input string could come from someone else so wouldn't an error be better?

I'm not sure why we elect to panic in `new` either FTR.

Reply:
We could add an issue and do it as a follow up if its not that important. I'm not super familiar with p2p

Reply:
@tcharding see the discussion above about exactly this.

We are not returning an error because every application is going to use a fixed string for this.